crates/provider/src/ext/trace.rs

```diff
@@ -178,18 +178,161 @@ where
+            let bob = &accounts[1];
+
+            let tx1 = TransactionRequest::default()
+                .with_from(*alice)
+                .with_to(*bob)
+                .with_nonce(0)
+                .with_value(U256::from(100));
+
+            let tx2 = TransactionRequest::default()
+                .with_from(*alice)
+                .with_to(*bob)
+                .with_nonce(1)
+                .with_value(U256::from(100));
+
+            let result = provider
+                .trace_call_many(&[(&tx1, &[TraceType::Trace]), (&tx2, &[TraceType::Trace])])
+                .await;
+            println!("{:?}", result);
+
+            // [
+            //     (
+            //         TransactionRequest {
+            //             from: Some(0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266),
+            //             to: Some(Call(0x70997970c51812dc3a010c7d01b50e0d17dc79c8)),
+            //             gas_price: None,
+            //             max_fee_per_gas: None,
+            //             max_priority_fee_per_gas: None,
+            //             max_fee_per_blob_gas: None,
+            //             gas: None,
+            //             value: Some(100),
+            //             input: TransactionInput { input: None, data: None },
+            //             nonce: Some(0),
+            //             chain_id: None,
+            //             access_list: None,
+            //             transaction_type: None,
+            //             blob_versioned_hashes: None,
+            //             sidecar: None,
+            //             authorization_list: None,
+            //         },
+            //         [Trace],
+            //     ),
+            //     (
+            //         TransactionRequest {
+            //             from: Some(0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266),
+            //             to: Some(Call(0x70997970c51812dc3a010c7d01b50e0d17dc79c8)),
+            //             gas_price: None,
+            //             max_fee_per_gas: None,
+            //             max_priority_fee_per_gas: None,
+            //             max_fee_per_blob_gas: None,
+            //             gas: None,
+            //             value: Some(100),
+            //             input: TransactionInput { input: None, data: None },
+            //             nonce: Some(1),
+            //             chain_id: None,
+            //             access_list: None,
+            //             transaction_type: None,
+            //             blob_versioned_hashes: None,
+            //             sidecar: None,
+            //             authorization_list: None,
+            //         },
+            //         [Trace],
+            //     ),
+            // ]
+
+            // Err(ErrorResp(ErrorPayload { code: -32602, message: "Invalid params", data:
+            // Some(RawValue("invalid type: map, expected a tuple of size 2 at line 1 column 1"))
+            // }))
+        })
+        .await
```

Comment:
@yash-atreya I'm stuck on a types / serialization issue here - mind having a look?

in Reth:

The signature:

```rust
    async fn trace_call_many(
        &self,
        calls: Vec<(TransactionRequest, HashSet<TraceType>)>,
        block_id: Option<BlockId>,
    ) -> RpcResult<Vec<TraceResults>>;
```

```rust
#[tokio::test(flavor = "multi_thread")]
async fn test_trace_call_many() {
    let handle = launch_http(vec![RethRpcModule::Trace]).await;
    let client = handle.http_client().unwrap();

    let x = TraceApiClient::trace_call_many(
        &client,
        vec![
            (TransactionRequest::default(), [TraceType::Trace].into()),
            (TransactionRequest::default(), [TraceType::Trace].into()),
        ],
        Some(BlockNumberOrTag::Latest.into()),
    )
    .await
    .unwrap();

    println!("{:?}", x);
}
```

yields:

```
[TraceResults { output: 0x, state_diff: None, trace: [TransactionTrace { action: Create(CreateAction { from: 0x0000000000000000000000000000000000000000, gas: 49947000, init: 0x, value: 0 }), error: None, result: Some(Create(CreateOutput { address: 0xbd770416a3345f91e4b34576cb804a576fa48eb1, code: 0x, gas_used: 0 })), subtraces: 0, trace_address: [] }], vm_trace: None }, TraceResults { output: 0x, state_diff: None, trace: [TransactionTrace { action: Create(CreateAction { from: 0x0000000000000000000000000000000000000000, gas: 49947000, init: 0x, value: 0 }), error: None, result: Some(Create(CreateOutput { address: 0x5a443704dd4b594b382c22a083e2bd3090a6fef3, code: 0x, gas_used: 0 })), subtraces: 0, trace_address: [] }], vm_trace: None }]
```

Confirming that the internals do work.

Full error is of this format:

```sh
2024-09-10T11:08:27.380824Z DEBUG jsonrpsee_core::proc_macros_support: Error parsing "calls" as "Vec < (TransactionRequest, HashSet < TraceType >) >": ErrorObject { code: InvalidParams, message: "Invalid params", data: Some(RawValue("invalid type: map, expected a tuple of size 2 at line 1 column 1")) }
```

Reply:
Turns out it was related to the formatting when passed to the `request` method not being a tuple (`(request,))` vs `request`)